File: backend-ts/src/llm/error-handling.ts
/**
 * LLM Error Handling — retry, fallback, and error classification.
 *
 * Ports backend/core/error_handling.py (327 LOC) to TypeScript.
 * Wraps pi-ai calls with resilience patterns:
 *   - Typed error hierarchy with retryable classification
 *   - Exponential backoff retry
 *   - Multi-provider fallback chain
 *   - Structured error logging
 */

// ---------------------------------------------------------------------------
// Error types
// ---------------------------------------------------------------------------

export enum ErrorType {
  API_ERROR = "api_error",
  TIMEOUT = "timeout",
  QUOTA_EXCEEDED = "quota_exceeded",
  AUTHENTICATION = "authentication",
  RATE_LIMIT = "rate_limit",
  NETWORK = "network",
  VALIDATION = "validation",
  UNKNOWN = "unknown",
}

export class LLMError extends Error {
  readonly errorType: ErrorType;
  readonly retryable: boolean;
  provider?: string;
  model?: string;
  readonly originalError?: unknown;

  constructor(
    message: string,
    errorType: ErrorType,
    retryable: boolean,
    opts?: { provider?: string; model?: string; originalError?: unknown },
  ) {
    super(message);
    this.name = "LLMError";
    this.errorType = errorType;
    this.retryable = retryable;
    this.provider = opts?.provider;
    this.model = opts?.model;
    this.originalError = opts?.originalError;
  }

  toDict(): Record<string, unknown> {
    return {
      error_type: this.errorType,
      message: this.message,
      retryable: this.retryable,
      provider: this.provider,
      model: this.model,
    };
  }
}

export class LLMAPIError extends LLMError {
  constructor(message: string, opts?: { provider?: string; model?: string; originalError?: unknown }) {
    super(message, ErrorType.API_ERROR, true, opts);
    this.name = "LLMAPIError";
  }
}

export class LLMTimeoutError extends LLMError {
  constructor(message: string, opts?: { provider?: string; model?: string; originalError?: unknown }) {
    super(message, ErrorType.TIMEOUT, true, opts);
    this.name = "LLMTimeoutError";
  }
}

export class LLMQuotaError extends LLMError {
  constructor(message: string, opts?: { provider?: string; model?: string; originalError?: unknown }) {
    super(message, ErrorType.QUOTA_EXCEEDED, false, opts);
    this.name = "LLMQuotaError";
  }
}

export class LLMAuthenticationError extends LLMError {
  constructor(message: string, opts?: { provider?: string; model?: string; originalError?: unknown }) {
    super(message, ErrorType.AUTHENTICATION, false, opts);
    this.name = "LLMAuthenticationError";
  }
}

export class LLMRateLimitError extends LLMError {
  constructor(message: string, opts?: { provider?: string; model?: string; originalError?: unknown }) {
    super(message, ErrorType.RATE_LIMIT, true, opts);
    this.name = "LLMRateLimitError";
  }
}

// ---------------------------------------------------------------------------
// Error classification
// ---------------------------------------------------------------------------

/**
 * Classify a raw error into a typed LLMError.
 * Uses string-based detection matching the Python implementation's keyword checks.
 */
export function classifyError(error: unknown, provider: string): LLMError {
  const errorStr = String(error).toLowerCase();
  const errorName = error instanceof Error ? error.constructor.name.toLowerCase() : "";

  const opts = { provider, originalError: error };

  // Timeout
  if (errorStr.includes("timeout") || errorStr.includes("timed out") || errorStr.includes("etimedout")) {
    return new LLMTimeoutError(`LLM request timed out (${provider})`, opts);
  }

  // Quota / billing
  if (
    errorStr.includes("quota") ||
    errorStr.includes("billing") ||
    errorStr.includes("insufficient_quota") ||
    errorStr.includes("overloaded")
  ) {
    return new LLMQuotaError(`LLM quota/billing issue (${provider}): ${errorStr.slice(0, 200)}`, opts);
  }

  // Authentication
  if (
    errorStr.includes("authentication") ||
    errorStr.includes("api_key") ||
    errorStr.includes("unauthorized") ||
    errorStr.includes("401") ||
    errorStr.includes("403") ||
    errorStr.includes("invalid_api_key")
  ) {
    return new LLMAuthenticationError(`LLM authentication failed (${provider})`, opts);
  }

  // Rate limit
  if (
    errorStr.includes("rate_limit") ||
    errorStr.includes("too_many_requests") ||
    errorStr.includes("429") ||
    errorStr.includes("rate limit")
  ) {
    return new LLMRateLimitError(`LLM rate limited (${provider})`, opts);
  }

  // Network / connection
  if (
    errorName.includes("connection") ||
    errorName.includes("http") ||
    errorName.includes("fetch") ||
    errorStr.includes("econnrefused") ||
    errorStr.includes("econnreset") ||
    errorStr.includes("enotfound") ||
    errorStr.includes("network")
  ) {
    return new LLMAPIError(`LLM network error (${provider}): ${errorStr.slice(0, 200)}`, opts);
  }

  // Default: generic API error (retryable)
  return new LLMAPIError(
    `LLM API error (${provider}): ${errorStr.slice(0, 200)}`,
    opts,
  );
}

// ---------------------------------------------------------------------------
// Structured logging
// ---------------------------------------------------------------------------

const log = {
  info: (msg: string) => console.info(`[llm] ${msg}`),
  warn: (msg: string) => console.warn(`[llm] WARN ${msg}`),
  error: (msg: string) => console.error(`[llm] ERROR ${msg}`),
};

export function structuredLogError(
  error: unknown,
  context: Record<string, unknown>,
): void {
  const logData: Record<string, unknown> = {
    error_type: error instanceof Error ? error.constructor.name : typeof error,
    error_message: String(error),
    ...context,
  };

  if (error instanceof LLMError) {
    logData.error_category = error.errorType;
    logData.retryable = error.retryable;
    logData.llm_provider = error.provider;
    logData.llm_model = error.model;
  }

  log.error(`Error occurred: ${JSON.stringify(logData)}`);
}

/**
 * Format user-friendly error message for agent responses.
 */
export function formatUserFriendlyError(error: unknown, agentName = "Agent"): string {
  if (error instanceof LLMError) {
    return `${agentName} encountered an issue: ${error.message}`;
  }
  return `${agentName} encountered an unexpected issue. Please try again or rephrase your request.`;
}

// ---------------------------------------------------------------------------
// Retry with exponential backoff
// ---------------------------------------------------------------------------

export interface RetryOptions {
  maxAttempts?: number;
  initialDelayMs?: number;
  backoffFactor?: number;
  retryableErrors?: Array<new (...args: never[]) => LLMError>;
}

const DEFAULT_RETRY_OPTIONS: Required<RetryOptions> = {
  maxAttempts: 3,
  initialDelayMs: 1000,
  backoffFactor: 2.0,
  retryableErrors: [LLMAPIError, LLMTimeoutError, LLMRateLimitError],
};

/**
 * Execute an async function with exponential backoff retry.
 *
 * Mirrors Python's @retry_with_exponential_backoff decorator.
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  opts?: RetryOptions,
): Promise<T> {
  const { maxAttempts, initialDelayMs, backoffFactor, retryableErrors } = {
    ...DEFAULT_RETRY_OPTIONS,
    ...opts,
  };

  let delay = initialDelayMs;

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (err) {
      const isRetryable =
        retryableErrors.some((cls) => err instanceof cls) ||
        (err instanceof LLMError && err.retryable);

      if (!isRetryable || attempt >= maxAttempts) {
        log.error(
          `Failed after ${attempt} attempt(s): ${err instanceof Error ? err.constructor.name : "Unknown"}: ${String(err).slice(0, 200)}`,
        );
        throw err;
      }

      log.warn(
        `Attempt ${attempt}/${maxAttempts}, retrying in ${delay}ms... ` +
          `Error: ${err instanceof Error ? err.constructor.name : "Unknown"}: ${String(err).slice(0, 200)}`,
      );

      await new Promise((resolve) => setTimeout(resolve, delay));
      delay *= backoffFactor;
    }
  }

  // Should not reach here, but TypeScript needs a return
  throw new LLMAPIError("Retry loop exhausted without success or throw");
}

// ---------------------------------------------------------------------------
// Fallback provider chain
// ---------------------------------------------------------------------------

/**
 * Static fallback chain — which providers to try when primary fails.
 * Mirrors Python's ErrorRecoveryStrategy.get_fallback_order().
 */
const FALLBACK_CHAINS: Record<string, string[]> = {
  anthropic: ["openai", "google", "groq", "ollama"],
  openai: ["anthropic", "google", "groq", "ollama"],
  google: ["anthropic", "openai", "groq", "ollama"],
  groq: ["anthropic", "openai", "google", "ollama"],
  ollama: ["anthropic", "openai", "google", "groq"],
};

/**
 * Get the ordered list of fallback providers for a given primary provider.
 */
export function getFallbackOrder(provider: string): string[] {
  return FALLBACK_CHAINS[provider] ?? Object.keys(FALLBACK_CHAINS).filter((p) => p !== provider);
}

/**
 * Determine whether we should try a different provider based on the error type.
 * Non-retryable errors (quota, auth) and persistent API errors warrant fallback.
 */
export function shouldFallbackToProvider(error: LLMError): boolean {
  return [ErrorType.QUOTA_EXCEEDED, ErrorType.AUTHENTICATION, ErrorType.API_ERROR].includes(
    error.errorType,
  );
}

File: backend-python-archive/core/error_handling.py
"""
Error handling utilities and middleware for RezNet AI
Provides robust error handling, retry logic, and user-friendly error messages
"""

import logging
import time
import functools
from typing import Optional, Dict, Any, Callable
from enum import Enum

logger = logging.getLogger(__name__)


class ErrorType(Enum):
    """Error types for categorization and handling"""
    API_ERROR = "api_error"
    TIMEOUT = "timeout"
    QUOTA_EXCEEDED = "quota_exceeded"
    AUTHENTICATION = "authentication"
    RATE_LIMIT = "rate_limit"
    NETWORK = "network"
    VALIDATION = "validation"
    UNKNOWN = "unknown"


class LLMError(Exception):
    """Base exception for LLM-related errors"""
    def __init__(
        self,
        message: str,
        error_type: ErrorType = ErrorType.UNKNOWN,
        retryable: bool = False,
        provider: Optional[str] = None,
        model: Optional[str] = None,
        original_error: Optional[Exception] = None
    ):
        super().__init__(message)
        self.message = message
        self.error_type = error_type
        self.retryable = retryable
        self.provider = provider
        self.model = model
        self.original_error = original_error

    def to_dict(self) -> Dict[str, Any]:
        """Convert error to dictionary for serialization"""
        return {
            "message": self.message,
            "error_type": self.error_type.value,
            "retryable": self.retryable,
            "provider": self.provider,
            "model": self.model
        }


class LLMAPIError(LLMError):
    """LLM API request failed"""
    def __init__(self, message: str, **kwargs):
        super().__init__(message, error_type=ErrorType.API_ERROR, retryable=True, **kwargs)


class LLMTimeoutError(LLMError):
    """LLM API request timed out"""
    def __init__(self, message: str, **kwargs):
        super().__init__(message, error_type=ErrorType.TIMEOUT, retryable=True, **kwargs)


class LLMQuotaError(LLMError):
    """LLM API quota exceeded"""
    def __init__(self, message: str, **kwargs):
        super().__init__(message, error_type=ErrorType.QUOTA_EXCEEDED, retryable=False, **kwargs)


class LLMAuthenticationError(LLMError):
    """LLM API authentication failed"""
    def __init__(self, message: str, **kwargs):
        super().__init__(message, error_type=ErrorType.AUTHENTICATION, retryable=False, **kwargs)


class LLMRateLimitError(LLMError):
    """LLM API rate limit exceeded"""
    def __init__(self, message: str, **kwargs):
        super().__init__(message, error_type=ErrorType.RATE_LIMIT, retryable=True, **kwargs)


def classify_error(error: Exception, provider: str) -> LLMError:
    """
    Classify a raw exception into a specific LLMError type

    Args:
        error: The original exception
        provider: LLM provider name (anthropic, openai, ollama)

    Returns:
        Classified LLMError with appropriate type and retryability
    """
    error_str = str(error).lower()
    error_type_name = type(error).__name__.lower()

    # Timeout errors
    if "timeout" in error_str or "timed out" in error_str:
        return LLMTimeoutError(
            f"Request timed out. The {provider} service is taking too long to respond.",
            provider=provider,
            original_error=error
        )

    # Quota/billing errors
    if any(keyword in error_str for keyword in ["quota", "billing", "insufficient_quota", "overloaded"]):
        return LLMQuotaError(
            f"Quota exceeded for {provider}. Please check your API usage limits.",
            provider=provider,
            original_error=error
        )

    # Authentication errors
    if any(keyword in error_str for keyword in ["authentication", "api_key", "unauthorized", "401", "403"]):
        return LLMAuthenticationError(
            f"Authentication failed for {provider}. Please check your API key configuration.",
            provider=provider,
            original_error=error
        )

    # Rate limit errors
    if any(keyword in error_str for keyword in ["rate_limit", "too_many_requests", "429"]):
        return LLMRateLimitError(
            f"Rate limit exceeded for {provider}. Please wait a moment and try again.",
            provider=provider,
            original_error=error
        )

    # Network/connection errors
    if any(keyword in error_type_name for keyword in ["connection", "http", "request"]):
        return LLMAPIError(
            f"Network error connecting to {provider}. Please check your internet connection.",
            provider=provider,
            original_error=error
        )

    # Generic API error (retryable)
    return LLMAPIError(
        f"An error occurred with the {provider} service. Retrying...",
        provider=provider,
        original_error=error
    )


def format_user_friendly_error(error: Exception, agent_name: str = "Agent") -> str:
    """
    Convert an exception into a user-friendly error message
    NEVER shows stack traces to end users

    Args:
        error: The exception to format
        agent_name: Name of the agent that encountered the error

    Returns:
        User-friendly error message suitable for displaying in chat
    """
    if isinstance(error, LLMError):
        # Use pre-formatted message from LLMError
        return f"⚠️ {agent_name} encountered an issue: {error.message}"

    # Generic error - hide technical details
    return f"⚠️ {agent_name} encountered an unexpected issue. Our team has been notified. Please try again or rephrase your request."


def structured_log_error(
    error: Exception,
    context: Dict[str, Any],
    level: str = "error"
) -> None:
    """
    Log error with full structured context for debugging

    Args:
        error: The exception to log
        context: Additional context (agent, model, provider, request, etc.)
        level: Log level (error, warning, info)
    """
    log_data = {
        "error_type": type(error).__name__,
        "error_message": str(error),
        **context
    }

    # Add LLMError-specific fields if applicable
    if isinstance(error, LLMError):
        log_data.update({
            "error_category": error.error_type.value,
            "retryable": error.retryable,
            "llm_provider": error.provider,
            "llm_model": error.model
        })

    # Log at appropriate level
    if level == "error":
        logger.error(f"Error occurred: {log_data}")
    elif level == "warning":
        logger.warning(f"Warning: {log_data}")
    else:
        logger.info(f"Info: {log_data}")


def retry_with_exponential_backoff(
    max_attempts: int = 3,
    initial_delay: float = 1.0,
    backoff_factor: float = 2.0,
    retryable_errors: tuple = (LLMAPIError, LLMTimeoutError, LLMRateLimitError)
):
    """
    Decorator to retry async functions with exponential backoff

    Args:
        max_attempts: Maximum number of retry attempts (default: 3)
        initial_delay: Initial delay in seconds (default: 1.0)
        backoff_factor: Multiplier for delay after each attempt (default: 2.0)
        retryable_errors: Tuple of exception types that should trigger retry

    Usage:
        @retry_with_exponential_backoff(max_attempts=3, initial_delay=1.0)
        async def my_function():
            # ... code that might fail
    """
    def decorator(func: Callable):
        @functools.wraps(func)
        async def wrapper(*args, **kwargs):
            delay = initial_delay
            last_error = None

            for attempt in range(1, max_attempts + 1):
                try:
                    return await func(*args, **kwargs)
                except Exception as e:
                    last_error = e

                    # Check if error is retryable
                    is_retryable = isinstance(e, retryable_errors) or (
                        isinstance(e, LLMError) and e.retryable
                    )

                    if not is_retryable or attempt >= max_attempts:
                        # Non-retryable error or max attempts reached
                        logger.error(
                            f"{func.__name__} failed after {attempt} attempt(s): "
                            f"{type(e).__name__}: {str(e)}"
                        )
                        raise

                    # Log retry attempt
                    logger.warning(
                        f"{func.__name__} failed on attempt {attempt}/{max_attempts}, "
                        f"retrying in {delay}s... Error: {type(e).__name__}: {str(e)}"
                    )

                    # Wait before retrying
                    time.sleep(delay)
                    delay *= backoff_factor

            # Should never reach here, but raise last error if we do
            if last_error:
                raise last_error

        return wrapper
    return decorator


class ErrorRecoveryStrategy:
    """
    Defines strategies for recovering from errors
    """

    @staticmethod
    def should_fallback_to_different_provider(error: LLMError) -> bool:
        """
        Determine if we should try a different LLM provider

        Args:
            error: The LLMError that occurred

        Returns:
            True if fallback to different provider should be attempted
        """
        # Fallback on quota, authentication, or persistent API errors
        return error.error_type in [
            ErrorType.QUOTA_EXCEEDED,
            ErrorType.AUTHENTICATION,
            ErrorType.API_ERROR
        ]

    @staticmethod
    def get_fallback_order(current_provider: str) -> list:
        """
        Get the order of providers to try as fallback

        Args:
            current_provider: The provider that failed

        Returns:
            List of provider names to try in order
        """
        # Define fallback order
        fallback_chains = {
            "anthropic": ["openai", "ollama"],
            "openai": ["anthropic", "ollama"],
            "ollama": ["anthropic", "openai"]
        }

        return fallback_chains.get(current_provider, [])

    @staticmethod
    def should_notify_user(error: LLMError) -> bool:
        """
        Determine if user should be notified about error

        Args:
            error: The LLMError that occurred

        Returns:
            True if user should see an error message
        """
        # Always notify user for non-retryable errors
        return not error.retryable or error.error_type in [
            ErrorType.AUTHENTICATION,
            ErrorType.QUOTA_EXCEEDED
        ]
